File: node/dynamic/gridTraveler.js
const gridTraveler = (n, m) => {
  if (m === 1 && n === 1) return 1;
  if (m === 0 || n === 0) return 0;

  return gridTraveler(n - 1, m) + gridTraveler(m - 1, n);
};

describe("", () => {
  it("", () => {
    expect(gridTraveler(1, 1)).toEqual(1);
    expect(gridTraveler(1, 2)).toEqual(1);
    expect(gridTraveler(2, 2)).toEqual(2);
    expect(gridTraveler(2, 3)).toEqual(3);
    expect(gridTraveler(2, 4)).toEqual(4);
    expect(gridTraveler(3, 3)).toEqual(6);
    expect(gridTraveler(3, 4)).toEqual(10);
  });
});

File: ruby/leetcode/grid_traveler.rb
def grid_traveler(i, j)
  paths = Array.new(i) { Array.new(j, 1) }
  paths.each_with_index do |p, i|
    next if i == 0

    p.each_with_index do |e, j|
      next if j == 0
      paths[i][j] = paths[i - 1][j] + paths[i][j - 1]
    end
  end

  paths[-1][-1]
end

describe "grid_traveler" do
  it do
    expect(grid_traveler(3, 7)).to eq(28)
  end
end
